### PR TITLE
Added token cache map to fix use of unintended auth token for subsequent connections

### PR DESCRIPTION
**Pull request by tkyc**
https://github.com/microsoft/mssql-jdbc/issues/2334

### CODE REVIEW

File: src/main/java/com/microsoft/sqlserver/jdbc/PersistentTokenCacheAccessAspect.java [hunk 1]
```diff
@@ -22,13 +22,16 @@
  * @see <a href="https://aka.ms/msal4j-token-cache">https://aka.ms/msal4j-token-cache</a>
  */
 public class PersistentTokenCacheAccessAspect implements ITokenCacheAccessAspect {
-    private static PersistentTokenCacheAccessAspect instance = new PersistentTokenCacheAccessAspect();
-
+    private static PersistentTokenCacheAccessAspect instance;
     private final Lock lock = new ReentrantLock();
 
-    private PersistentTokenCacheAccessAspect() {}
+    static final long TIME_TO_LIVE = 86400000L; // Token cache time to live (24 hrs).
```
Comment (barryw-mssql) [inline src/main/java/com/microsoft/sqlserver/jdbc/PersistentTokenCacheAccessAspect.java:28]: "Magic Number". Please use a numeric constant variable with a descriptive name instead of a numeric literal. You should never need to add a comment to document the value/meaning/purpose of a numeric literal. If this is done for all numeric values it will make the code more self documenting and simplify future maintenance efforts

File: src/test/java/com/microsoft/sqlserver/jdbc/TestResource.java [hunk 2]
```diff
@@ -63,6 +63,11 @@ protected Object[][] getContents() {
             {"R_ConnectionURLNull", "The connection URL is null."},
             {"R_connectionIsNotClosed", "The connection is not closed."},
             {"R_invalidExceptionMessage", "Invalid exception message"},
+            {"R_invalidClientSecret", "AADSTS7000215: Invalid client secret provided"},
+            {"R_invalidCertFields",
+                    "Error reading certificate, please verify the location of the certificate.signed fields invalid"},
```
Comment (barryw-mssql) [inline src/test/java/com/microsoft/sqlserver/jdbc/TestResource.java:68]: Is "certificate.signed" a typo?

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java [hunk 3]
```diff
@@ -53,6 +55,7 @@
 
 import com.microsoft.sqlserver.jdbc.SQLServerConnection.ActiveDirectoryAuthentication;
 import com.microsoft.sqlserver.jdbc.SQLServerConnection.SqlFedAuthInfo;
+import mssql.googlecode.concurrentlinkedhashmap.ConcurrentLinkedHashMap;
```
Comment (Jeffery-Wasty) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java:58]: Unused.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java [hunk 4]
```diff
@@ -84,10 +89,17 @@ static SqlAuthenticationToken getSqlFedAuthToken(SqlFedAuthInfo fedAuthInfo, Str
         lock.lock();
 
         try {
+            String hashedSecret = getHashedSecret(new String[] {fedAuthInfo.stsurl, user, password});
+            PersistentTokenCacheAccessAspect persistentTokenCacheAccessAspect;
+
+            if (null == (persistentTokenCacheAccessAspect = tokenCacheMap.getEntry(hashedSecret))) {
+                persistentTokenCacheAccessAspect = new PersistentTokenCacheAccessAspect();
+                tokenCacheMap.addEntry(hashedSecret, persistentTokenCacheAccessAspect);
+            }
```
Comment (Jeffery-Wasty) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java:98]: Having assignment and comparison together is a bit hard to parse. I'd prefer this.
```suggestion
            PersistentTokenCacheAccessAspect persistentTokenCacheAccessAspect = tokenCacheMap.getEntry(hashedSecret);

            if (null == persistentTokenCacheAccessAspect) {
                persistentTokenCacheAccessAspect = new PersistentTokenCacheAccessAspect();
                tokenCacheMap.addEntry(hashedSecret, persistentTokenCacheAccessAspect);
            }
```

File: src/test/java/com/microsoft/sqlserver/jdbc/TestResource.java (hunk 2 above)
Comment (tkyc) [reply src/test/java/com/microsoft/sqlserver/jdbc/TestResource.java:68]: Nope not a typo. That's the exact error message to check for.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java [hunk 5]
```diff
@@ -132,11 +144,19 @@ static SqlAuthenticationToken getSqlFedAuthTokenPrincipal(SqlFedAuthInfo fedAuth
...
 
         try {
+            String hashedSecret = getHashedSecret(
+                    new String[] {fedAuthInfo.stsurl, aadPrincipalID, aadPrincipalSecret});
+            PersistentTokenCacheAccessAspect persistentTokenCacheAccessAspect;
+
+            if (null == (persistentTokenCacheAccessAspect = tokenCacheMap.getEntry(hashedSecret))) {
+                persistentTokenCacheAccessAspect = new PersistentTokenCacheAccessAspect();
+                tokenCacheMap.addEntry(hashedSecret, persistentTokenCacheAccessAspect);
+            }
```
Comment (Jeffery-Wasty) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java:154]: Same as above.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java [hunk 6]
```diff
@@ -181,6 +201,15 @@ static SqlAuthenticationToken getSqlFedAuthTokenPrincipalCertificate(SqlFedAuthI
...
 
         try {
+            String hashedSecret = getHashedSecret(new String[] {fedAuthInfo.stsurl, aadPrincipalID, certFile,
+                    certPassword, certKey, certKeyPassword});
+            PersistentTokenCacheAccessAspect persistentTokenCacheAccessAspect;
+
+            if (null == (persistentTokenCacheAccessAspect = tokenCacheMap.getEntry(hashedSecret))) {
+                persistentTokenCacheAccessAspect = new PersistentTokenCacheAccessAspect();
+                tokenCacheMap.addEntry(hashedSecret, persistentTokenCacheAccessAspect);
+            }
```
Comment (Jeffery-Wasty) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java:211]: Ditto

File: src/main/java/com/microsoft/sqlserver/jdbc/PersistentTokenCacheAccessAspect.java (hunk 1 above)
Comment (tkyc) [reply src/main/java/com/microsoft/sqlserver/jdbc/PersistentTokenCacheAccessAspect.java:28]: `TIME_TO_LIVE` is the numeric constant. I can be more descriptive in the name eg. maybe `TWENTY_FOUR_...` but that gets long and potentially `TIME_TO_LIVE`'s value can change. So, a compromise would be to use a short name and a comment indicating the time in hours (b/c 86400000 ms isn't immediately obvious that it's 24 hours) if someone were to ever directly look at the constant.

File: src/test/java/com/microsoft/sqlserver/jdbc/TestResource.java [hunk 7]
```diff
@@ -63,6 +63,11 @@ protected Object[][] getContents() {
...
+            {"R_invalidAADPasswordAuth",
+                    "Failed to authenticate the user {0} in Active Directory (Authentication=ActiveDirectoryPassword)"},
```
Comment (Jeffery-Wasty) [inline src/test/java/com/microsoft/sqlserver/jdbc/TestResource.java:70]: Maybe future-proof this and make it like R_MSALExecution in SQLServerTestResource

```suggestion
            {"R_invalidAADAuth",
                    "Failed to authenticate the user {0} in Active Directory (Authentication={1})."},
```

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java [hunk 8]
```diff
@@ -449,4 +476,47 @@ private static SQLServerException getCorrectedException(Exception e, String user
...
+                if (System.currentTimeMillis() > persistentTokenCacheAccessAspect.getExpiryTime()) {
+                    tokenCacheMap.remove(key);
+
+                    persistentTokenCacheAccessAspect = new PersistentTokenCacheAccessAspect();
+                    persistentTokenCacheAccessAspect
+                            .setExpiryTime(System.currentTimeMillis() + PersistentTokenCacheAccessAspect.TIME_TO_LIVE);
+
+                    tokenCacheMap.put(key, persistentTokenCacheAccessAspect);
+
+                    return persistentTokenCacheAccessAspect;
```
Comment (Jeffery-Wasty) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java:510]: Super minor, but this can be removed since we're immediately returning below.

File: src/test/java/com/microsoft/sqlserver/jdbc/fedauth/FedauthTest.java [hunk 9]
```diff
@@ -16,8 +16,10 @@
 import java.security.cert.CertificateFactory;
 import java.sql.Connection;
 import java.sql.DriverManager;
+import java.sql.ResultSet;
```
Comment (Jeffery-Wasty) [inline src/test/java/com/microsoft/sqlserver/jdbc/fedauth/FedauthTest.java:19]: Leftover from a prior commit, unused.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java [hunk 10]
```diff
@@ -61,6 +63,8 @@ class SQLServerMSAL4JUtils {
     static final String SLASH_DEFAULT = "/.default";
     static final String ACCESS_TOKEN_EXPIRE = "access token expires: ";
 
+    private static final TokenCacheMap tokenCacheMap = new TokenCacheMap();
```
Comment (Jeffery-Wasty) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java:66]: I missed this, but rename to `TOKEN_CACHE_MAP` as its `final` `static`.

**Review by barryw-mssql — APPROVED** [review]
